jwst/outlier_detection/utils.py

```diff
@@ -20,177 +26,365 @@
 _ONE_MB = 1 << 20
 
 
+def nanmedian3D(cube, overwrite_input=True):
+    """Compute the median of a cube ignoring warnings and with memory efficiency.
+    np.nanmedian always uses at least 64-bit precision internally, and this is too
+    memory-intensive. Instead, loop over the median calculation to avoid the
+    memory usage of the internal upcasting and temporary array allocation.
+    The additional runtime of this loop is indistinguishable from zero,
+    but this loop cuts overall step memory usage roughly in half for at least one
+    test association.
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings(action="ignore",
+                                message="All-NaN slice encountered",
+                                category=RuntimeWarning)
+        output_arr = np.empty(cube.shape[1:], dtype=np.float32)
+        for i in range(output_arr.shape[0]):
+            # this for loop looks silly, but see docstring above
+            output_arr[i,:] = np.nanmedian(cube[:,i,:], axis=0, overwrite_input=overwrite_input).astype(np.float32)
+        return output_arr
+
+
 def create_cube_median(cube_model, maskpt):
     log.info("Computing median")
 
     weight_threshold = compute_weight_threshold(cube_model.wht, maskpt)
 
-    median = np.nanmedian(
+    # not safe to use overwrite_input=True here because we are operating on model.data directly
+    return nanmedian3D(
         np.ma.masked_array(cube_model.data, np.less(cube_model.wht, weight_threshold), fill_value=np.nan),
-        axis=0,
-    )
-    return median
+        overwrite_input=False)
 
 
-def create_median(resampled_models, maskpt, buffer_size=10.0):
-    """Create a median image from the singly resampled images.
+def drizzle_and_median(input_models,
+                       resamp,
+                       maskpt,
+                       resample_data=False,
+                       save_intermediate_results=False,
+                       make_output_path=None,
+                       buffer_size=None):
+    """
+    Shared code between imaging and spec modes for resampling and median computation
 
     Parameters
     ----------
-    resampled_models : ModelLibrary
-        The singly resampled images.
+    input_models : ModelLibrary
+        The input datamodels.
+
+    resamp : resample.resample.ResampleData object
+        The controlling object for the resampling process.
```

Comment:
I find this a bit confusing and am wondering if there's a different way to refactor this to avoid some of the confusion. `resamp` is a `ResampleData` instance passed into this function that is used for:
- resampling (only if `resample_data=True`)
- weight type and good bits (if `resample_data=False`)
- output_dir and asn_id (if `save_intermediate_results=True`)

`resamp.output_wcs` is only used when `resample_data=True` (but is computed in both cases since it's computed in the constructor for `ResampleData`).

Can to `resample=False` uses of `resamp` be removed (and replaced with maybe uses of `weight_type`, `good_bits` and patching `make_output_path` to handle both the `output_dir` and `asn_id`)? That way for `resample=False` there's no need to make a `ResampleData` instance.

Reply:
I did the suggested refactor and a bit more.  Let me know if you like the new setup better.  I wasn't sure how best to avoid repeating code, and I'm a little bit suspicious of some of the choices I made...